### PR TITLE
Fix race condition in Global Distribution packet retransmission

### LINKED ISSUE #1683: Disable broken test_in_order_messages_on_multiple_connections_with_bounce test case in GlobDist
MongooseIM version: 4a19c96177d50737380ea552f873aadf1c90136b

Broken test:
http://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1680/4000/pgsql_mnesia.20.0/big/ct_run.test@travis-job-c416160e-42d3-4434-ade5-ca16966418d8.2018-01-22_16.46.45/ejabberd_tests.tests.mod_global_distrib_SUITE.logs/run.2018-01-22_17.02.33/mod_global_distrib_suite.test_in_order_messages_on_multiple_connections_with_bounce.html

from mod_global_distrib_server_mgr

```erlang
-spec pick_connection(Enabled :: [endpoint_info()]) -> pid().                        
pick_connection(Enabled) ->                                                          
    #endpoint_info{ conn_pool_ref = PoolRef } = lists:nth(rand:uniform(length(Enabled)), Enabled),
    cpool:get_connection(PoolRef). 
```

This is not correct. Picking random connection would not enforce order.
There are several ways to do it right:
- use hash from `From` jid;
- use hash from `{From, To}`;
- use hash of the caller's pid. i.e. the pid of the process, sending stanza, or calling routing function.

### PR DESCRIPTION
**Pull request by fenek**
This PR fixes #1683 and obsoletes #1684

The race condition was provoked by `test_in_order_messages_on_multiple_connections_with_bounce` test case. It manifested under following conditions:

1. Due to lack of recipient mapping, e.g. 4 messages were stored in bounce storage
2. Mapping appeared
3. 5th message got sent and it triggered `known_recipient` hook, which in turn triggered retransmission of stored items
4. For every retransmitted item, full routing logic is executed, so each time the recipient mapping is retrieved
5. Test process may remove (as a part of test case) recipient mapping in the middle of retransmission, so only some of the items reach recipient.
6. Code returns from hook execution to a point, where the mapping is already retrieved and the message 5 gets delivered anyway
7. Result: messages delivered out of order

I've decided not to create new test case nor make the existing one more predictable, as with 100 messages sent it reproduces the issue fairly frequently and there is very little chance it won't test what it is supposed to. I mean an edge case when e.g. all 100 messages hit a window where recipient mapping is available due to test-c2s concurrency.

Note: This version was built on Travis about 6 times with full set of jobs and `test_in_order_messages_on_multiple_connections_with_bounce` was never a reason for crash.

### CODE REVIEW

**Comment by arcusfelis** [comment]
pamparam, don't like any concurrency in this quantum world^W^W^W any tests.

File: src/global_distrib/mod_global_distrib.erl [hunk 1]
```diff
@@ -76,17 +76,20 @@ maybe_reroute({From, To, Acc0, Packet} = FPacket) ->
     Acc = maybe_initialize_metadata(Acc0),
     LocalHost = opt(local_host),
     GlobalHost = opt(global_host),
-    case lookup_recipients_host(To, LocalHost, GlobalHost) of
+    case lookup_recipients_host(get_metadata(Acc, target_host_override, undefined),
```
Comment (arcusfelis) [inline src/global_distrib/mod_global_distrib.erl]: Can we have it on a separate line?

```erlang
TargetHost = get_metadata(Acc, target_host_override, undefined),
```

It still will be two lines long :D

Comment (arcusfelis) [reply src/global_distrib/mod_global_distrib.erl]: target_host_override should be documented, I have no idea what is for just by looking at the code.

Comment (arcusfelis) [reply src/global_distrib/mod_global_distrib.erl]: and why it can be undefined should be described too.
